src/EditorFeatures/CSharp/CompleteStatement/CompleteStatementCommandHandler.cs

```diff
@@ -0,0 +1,397 @@
+        {
+            if (args.TypedChar != ';')
+            {
+                return;
+            }
+
+            var caret = args.TextView.GetCaretPoint(args.SubjectBuffer);
+            if (!caret.HasValue)
+            {
+                return;
+            }
+
+            var isCaretAtEndOfLine = ((SnapshotPoint)caret).Position == ((SnapshotPoint)caret).GetContainingLine().End;
+            var document = caret.Value.Snapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document == null)
+      
+            {
+                return;
+            }
+
+            var root = document.GetSyntaxRootSynchronously(executionContext.OperationContext.UserCancellationToken);
+            var caretPosition = caret.Value.Position;
+
+            var token = root.FindToken(caretPosition);
+
+            var currentNode = token.Parent;
+
+            // If cursor is right before an opening delimiter, start with node outside of delimiters. 
+            // This covers cases like `obj.ToString$()`, where `token` references `(` but the caret isn't actually 
+            // inside the argument list.
+            if (token.IsKind(SyntaxKind.OpenBraceToken, SyntaxKind.OpenBracketToken, SyntaxKind.OpenParenToken)
+                && token.Span.Start >= caretPosition)
+            {
+                currentNode = currentNode.Parent;
+            }
+
+            if (currentNode == null)
+            {
+                return;
+            }
+
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            if (!LooksLikeNodeInArgumentListForStatementCompletion(currentNode, syntaxFacts, isCaretAtEndOfLine))
+            {
+                return;
+            }
+
+            // verify all delimeters exist until you reach statement syntax that requires a semicolon
+            while (!IsStatementOrFieldDeclaration(currentNode, syntaxFacts))
+            {
+                if (!ClosingDelimiterExistsIfNeeded(currentNode))
+                {
+                    // A required delimiter is missing; do not treat semicolon as statement completion
+                    return;
+                }
+
+                if (currentNode.Parent == null)
+                {
+                    return;
+                }
+
+                currentNode = currentNode.Parent;
+            }
+
+            // if the statement syntax itself requires a closing delimeter, verify it is there
+                if (StatementClosingDelimiterIsMissing(currentNode))
+                {
+                    // Example: missing final `)` in `do { } while (x$$`
+                    return;
+                }
+
+            var semicolonPosition = GetSemicolonLocation(currentNode, caretPosition);
+
+            // Place cursor after the statement
+            args.TextView.TryMoveCaretToAndEnsureVisible(args.SubjectBuffer.CurrentSnapshot.GetPoint(GetEndPosition(root, semicolonPosition, currentNode.Kind())));
```

Comment:
wrap long line.

Reply:
➡️ Resolving as lower priority and covered by https://github.com/dotnet/roslyn/pull/30743#discussion_r242949053